internal/api/installation_db_migration.go

```diff
@@ -0,0 +1,325 @@
+	vars := mux.Vars(r)
+	migrationID := vars["migration"]
+
+	c.Logger = c.Logger.WithField("action", "commit-installation-database-migration").
+		WithField("migration-operation", migrationID)
+
+	dbMigrationOperation, status, unlockOnce := lockInstallationDBMigrationOperation(c, migrationID)
+	if status != 0 {
+		w.WriteHeader(status)
+		return
+	}
+	defer unlockOnce()
+
+	if dbMigrationOperation.State != model.InstallationDBMigrationStateSucceeded {
+		c.Logger.Warn("Cannot commit DB migration that hasn't succeeded")
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	sourceDB := c.DBProvider.GetDatabase(dbMigrationOperation.InstallationID, dbMigrationOperation.SourceDatabase)
+
+	err := sourceDB.TeardownMigrated(c.Store, dbMigrationOperation, c.Logger)
+	if err != nil {
+		c.Logger.WithError(err).Error("Failed to tear down migrated database")
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	dbMigrationOperation.State = model.InstallationDBMigrationStateCommitted
+	err = c.Store.UpdateInstallationDBMigrationOperationState(dbMigrationOperation)
+	if err != nil {
+		c.Logger.WithError(err).Error("Failed to set operation status to committed")
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	w.WriteHeader(http.StatusOK)
+	outputJSON(c, w, dbMigrationOperation)
+}
+
+// handleRollbackInstallationDatabaseMigration responds to POST /api/installations/operations/database/migration/{migration}/rollback,
+// rollbacks database migration.
+func handleRollbackInstallationDatabaseMigration(c *Context, w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+	migrationID := vars["migration"]
+
+	c.Logger = c.Logger.WithField("action", "rollback-installation-database-migration").
+		WithField("migration-operation", migrationID)
+
+	dbMigrationOperation, status, unlockOnce := lockInstallationDBMigrationOperation(c, migrationID)
+	if status != 0 {
+		w.WriteHeader(status)
+		return
+	}
+	defer unlockOnce()
+
+	newState := model.InstallationDBMigrationStateRollbackRequested
+
+	if !dbMigrationOperation.ValidTransitionState(newState) {
+		c.Logger.Warn("Cannot rollback migration, invalid state")
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	installationDTO, status, unlockInstOnce := lockInstallation(c, dbMigrationOperation.InstallationID)
+	if status != 0 {
+		w.WriteHeader(status)
+		return
+	}
+	defer unlockInstOnce()
+
+	if installationDTO.State != model.InstallationStateHibernating {
+		c.Logger.Error("Installation needs to be hibernated to be rolled back")
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	err := c.Store.TriggerInstallationDBMigrationRollback(dbMigrationOperation, installationDTO.Installation)
+	if err != nil {
+		c.Logger.WithError(err).Error("Failed to trigger db migration rollback")
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	unlockOnce()
+	c.Supervisor.Do()
```

Comment:
`Do` will send notification to scheduler's channel that will trigger supervisors to work and it will not receive new messages until they are done, therefore other simultanous call to `Do` will be ignored. You can see details here https://github.com/mattermost/mattermost-cloud/blob/master/internal/supervisor/scheduler.go#L38-L50.

We use this pattern in other API handlers as well https://github.com/mattermost/mattermost-cloud/blob/master/internal/api/installation.go#L231.